File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/ProvideDeserializationMethodsForOptionalFields.vb
Imports System
Imports System.Runtime.Serialization

Namespace Tests.Diagnostics
    Public Class NoEventHandlerMethods ' Noncompliant {{Add deserialization event handlers.}}
'                ^^^^^^^^^^^^^^^^^^^^^
        <OptionalField>
        Private optionalField As Integer = 5
    End Class

    Public Class OnlyOnDeserializingEventHandlerMethod '  Noncompliant {{Add the missing 'OnDeserializedAttribute' event handler.}}
        <OptionalFieldAttribute>
        Private optionalField As Integer = 5

        <OnDeserializing>
        Private Sub OnDeserializing(ByVal context As StreamingContext)
            optionalField = 5
        End Sub
    End Class

    Public Class OnlyOnDeserializedEventHandlerMethod ' Noncompliant {{Add the missing 'OnDeserializingAttribute' event handler.}}
        <OptionalField>
        Private optionalField As Integer = 5

        <OnDeserialized>
        Private Sub OnDeserialized(ByVal context As StreamingContext)
        End Sub
    End Class

    <Serializable>
    Public Class Compliant
        <OptionalField>
        Private optionalField As Integer = 5

        <OnDeserializing>
        Private Sub OnDeserializing(ByVal context As StreamingContext)
            optionalField = 5
        End Sub

        <OnDeserialized>
        Private Sub OnDeserialized(ByVal context As StreamingContext)
        End Sub
    End Class
End Namespace

File: analyzers/tests/SonarAnalyzer.Test/TestCases/ProvideDeserializationMethodsForOptionalFields.cs
using System;
using System.Runtime.Serialization;

namespace Tests.Diagnostics
{
    public class NoEventHandlerMethods // Noncompliant {{Add deserialization event handlers.}}
//               ^^^^^^^^^^^^^^^^^^^^^
    {
        [OptionalField]
        int optionalField = 5;
    }

    public class OnlyOnDeserializingEventHandlerMethod // Noncompliant {{Add the missing 'OnDeserializedAttribute' event handler.}}
    {
        [OptionalFieldAttribute]
        int optionalField = 5;

        [OnDeserializing]
        void OnDeserializing(StreamingContext context)
        {
            optionalField = 5;
        }
    }

    public class OnlyOnDeserializedEventHandlerMethod // Noncompliant {{Add the missing 'OnDeserializingAttribute' event handler.}}
    {
        [OptionalField]
        int optionalField = 5;

        [OnDeserialized]
        void OnDeserialized(StreamingContext context)
        {
            // Set optionalField if dependent on other deserialized values.
        }
    }

    [Serializable]
    public class Compliant
    {
        [OptionalField]
        int optionalField = 5;

        [OnDeserializing]
        void OnDeserializing(StreamingContext context)
        {
            optionalField = 5;
        }

        [OnDeserialized]
        void OnDeserialized(StreamingContext context)
        {
            // Set optionalField if dependent on other deserialized values.
        }
    }

    public struct NoncompliantStruct // Noncompliant {{Add deserialization event handlers.}}
    {
        [OptionalField]
        int optionalField;
    }
}
